apis/python/src/tiledbsoma/dense_nd_array.py

```diff
@@ -148,16 +151,64 @@ def read_tensor(
         """
-        tiledb_result_order = tiledb_result_order_from_soma_result_order(
-            result_order, accept=["column-major", "row-major"]
-        )
         with self._tiledb_open("r") as A:
             target_shape = _dense_index_to_shape(coords, A.shape, result_order)
-            query = A.query(return_arrow=True, order=tiledb_result_order)
-            arrow_tbl = query.df[coords]
-            return pa.Tensor.from_numpy(
-                arrow_tbl.column("soma_data").to_numpy().reshape(target_shape)
-            )
+            schema = A.schema
+
+        sr = clib.SOMAReader(
+            self._uri,
+            name=self.__class__.__name__,
+            result_order=result_order,
+            platform_config={} if self._ctx is None else self._ctx.config().dict(),
+        )
+
+        if coords is not None:
+            if not isinstance(coords, (list, tuple)):
+                raise TypeError(
+                    f"coords type {type(coords)} unsupported; expected list or tuple"
+                )
+            if len(coords) < 1 or len(coords) > schema.domain.ndim:
+                raise ValueError(
+                    f"coords {coords} must have length between 1 and ndim ({schema.domain.ndim}); got {len(coords)}"
+                )
+
+            for i, coord in enumerate(coords):
+                #                # Example: coords = [None, 3, slice(4,5)]
+                #                # coor takes on values None, 3, and slice(4,5) in this loop body.
+                dim_name = schema.domain.dim(i).name
+                if coord is None:
+                    pass  # No constraint; select all in this dimension
+                elif isinstance(coord, int):
+                    sr.set_dim_points(dim_name, [coord])
+                elif isinstance(coord, slice):
+                    lo_hi = util.slice_to_range(coord)
+                    if lo_hi is not None:
+                        lo, hi = lo_hi
+                        if lo < 0 or hi < 0:
+                            raise ValueError(
+                                f"slice start and stop may not be negative; got ({lo}, {hi})"
+                            )
+                        if lo > hi:
+                            raise ValueError(
+                                f"slice start must be <= slice stop; got ({lo}, {hi})"
+                            )
+                        sr.set_dim_ranges(dim_name, [lo_hi])
+                    # Else, no constraint in this slot. This is `slice(None)` which is like
+                    # Python indexing syntax `[:]`.
+                elif isinstance(
+                    coord, (collections.abc.Sequence, pa.Array, pa.ChunkedArray)
+                ):
+                    sr.set_dim_points(dim_name, coord)
+                else:
+                    raise TypeError(f"coord type {type(coord)} at slot {i} unsupported")
+
+        sr.submit()
+
+        arrow_table = sr.read_next()
+        # XXX assert no more data remaining
```

Comment:
@bkmartinjr : in the spec we don't have a generator for the read in `DenseNdArray`.

@gspowley : what's your advice on how to handle ensuring we really do have all the data?

Reply:
You can call `sr.results_complete()`. It returns `true` if the last call to `sr.read_next()` returned all results for the query.

Reply:
And FYI a working (if minimal) implementation of `setup; while !complete: read next;` is in the R package [in this file](https://github.com/single-cell-data/TileDB-SOMA/blob/main/apis/r/src/riterator.cpp).

Reply:
> in the spec we don't have a generator for the read in DenseNdArray

Correct - we decided to remove it as:
* we don't support point indexing
* the user can determine a priori the size of the result
* and therefore do their own partitioning/iteration 

We also agreed that we would revisit this if it ever became onerous

Reply:
And I think you are implicitly acknowledging it in your comment, but this code does not appear to work correctly for reads that exceed the soma buffer size.  it needs do the iteration, concat the tables, and then create the tensor.

@gspowley - I believe using tables in this way results in an extra copy as you can't create a tensor from a table with zero copy.  At some point in the future, we would benefit from a "read-as-tensor" C++ interface.